### PR TITLE
Nse solver

### PR DESCRIPTION
**Pull request by zhichen3**
PR for implementing nse solver as a replacement for pr #960 due to unknown issues.

### CODE REVIEW

File: networks/ignition_reaclib/URCA-simple/reaclib_rate_metadata.dat [hunk 1]
```diff
@@ -0,0 +1,2 @@
+<secret_code>(0)
```
Comment (zingale) [inline networks/ignition_reaclib/URCA-simple/reaclib_rate_metadata.dat:1]: this looks like it was added by mistake?

File: networks/ignition_reaclib/URCA-simple/reaclib_rates.H [hunk 2]
```diff
@@ -220,6 +221,7 @@ fill_reaclib_rates(const tf_t& tfactors, rate_eval_t& rate_eval)
     rate_eval.dscreened_rates_dT(k_n__p__weak__wc12) = drate_dT;
 
 
+
```
Comment (zingale) [inline networks/ignition_reaclib/URCA-simple/reaclib_rates.H:224]: this file should be automatically generated -- did you mean to update it here?

File: networks/ignition_reaclib/URCA-simple/table_rates_data.cpp [hunk 3]
```diff
@@ -33,18 +33,16 @@ void init_tabular()
     j_na23_ne23_meta.nrhoy = 152;
     j_na23_ne23_meta.nvars = 6;
     j_na23_ne23_meta.nheader = 6;
-    j_na23_ne23_meta.file = "23Na-23Ne_electroncapture.dat";
 
-    init_tab_info(j_na23_ne23_meta, j_na23_ne23_rhoy, j_na23_ne23_temp, j_na23_ne23_data);
+    init_tab_info(j_na23_ne23_meta, "23Na-23Ne_electroncapture.dat", j_na23_ne23_rhoy, j_na23_ne23_temp, j_na23_ne23_data);
```
Comment (zingale) [inline networks/ignition_reaclib/URCA-simple/table_rates_data.cpp:37]: likewise here -- it seems that the update of URCA-sinple is independent of this PR, so maybe revert these files back to the old version?

File: nse/actual_nse.H [hunk 4]
```diff
@@ -0,0 +1,214 @@
...
+  amrex::Array2D<amrex::Real, 0, 1, 0, 1> jac;
+};
+
+template <typename T>
+T get_nse_state(T& state)
+{
+  // This function finds the nse state given the burn state or eos state
+
+  // three constants for calculating coulomb correction term
+  const amrex::Real A1 = -0.9052_rt;
```
Comment (zingale) [inline nse/actual_nse.H:28]: can  you add some comments telling us where these constants come from? and what their units are?

File: nse/actual_nse.H [hunk 5]
```diff
@@ -0,0 +1,214 @@
...
+  // Need partition function, set it to 1 for now.
+  amrex::Real partition_function = 1.0_rt;
+
+  for (int n = 0; n < NumSpec; ++n){
+    if (short_spec_names_cxx[n] == "p"){
+      continue;
+    }
+    // term for calculating u_c
+    gamma = std::pow(zion[n], 5.0_rt/3.0_rt) * C::q_e * C::q_e *
+      std::pow(4.0_rt * M_PI * n_e / 3.0_rt, 1.0_rt / 3.0_rt) / C::k_B / state.T;
```
Comment (zingale) [inline nse/actual_nse.H:49]: you can use `std::cbrt()` here

**Comment by zhichen3** [comment]
yah I didn't mean to change anything related to URCA stuff, its probably because I merged main and weird stuff happened.

File: nse/actual_nse.H [hunk 6]
```diff
@@ -0,0 +1,214 @@
...
+
+  return nse_inputs;
+}
+
+// A newton-raphson solver for finding nse state
+// used for calibrating chemical potential of proton and neutron
+template<typename T>
+void nse_nr_solver(T& state, amrex::Real eps=1.0e-3_rt, int nloops=500){
+  
+  // Check if network results in singular jacobian first, e.g. aprox13 and iso7
```
Comment (zingale) [inline nse/actual_nse.H:121]: can you expand this more -- what makes a network singular? and why are aprox13 and iso7 special here?

File: unit_test/test_nse/README.md [hunk 7]
```diff
@@ -0,0 +1,17 @@
...
+`burn_cell` integrates a network for a single set of initial
+conditions.  The density, temperature, and composition are set in the
+inputs file, as well as the maximum time to integrate.
+
+Also calculates the nse mass fraction before and after integration.
+
+Upon completion, the new state is printed to the screen.
+
+
+## continuous integration
```
Comment (zingale) [inline unit_test/test_nse/README.md:12]: I don't think this comment applies here -- can you update the README?

File: unit_test/test_nse/inputs_sn160 [hunk 8]
```diff
@@ -0,0 +1,189 @@
...
+unit_test.X151 = 0.0001
+unit_test.X152 = 0.0
+unit_test.X153 = 0.0
+unit_test.X154 = 0.0
+unit_test.X155 = 0.0
+unit_test.X156 = 0.0
+unit_test.X157 = 0.0
+unit_test.X158 = 0.0
+unit_test.X159 = 0.0
+unit_test.X160 = 0.0
```
Comment (zingale) [inline unit_test/test_nse/inputs_sn160:186]: I don't think this inputs makes sense -- we don't permit this many mass fractions to be input (see _parameters)

File: unit_test/test_nse/matplotlibrc [hunk 9]
```diff
@@ -0,0 +1,7 @@
+font.family        : serif
+font.serif         : Computer Modern Roman, Times, Palatino, New Century Schoolbook, Bookman
```
Comment (zingale) [inline unit_test/test_nse/matplotlibrc:2]: I don't think we should include a matplotlibrc in a single test -- we haven't done so anywhere else

**Comment by zhichen3** [comment]
okay, I think I addressed all the comments in the new push

**Review by zingale — APPROVED** [review]

**Pull request merged**